File: src/main/java/五毒/第一周/第二遍/threeSum.java
package 五毒.第一周.第二遍;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

public class three_Sum {
    public List<List<Integer>> threeSum(int[] nums) {
        if(nums.length==0){
            return new ArrayList<>();
        }
        Arrays.sort(nums);
        List<List<Integer>> result = new ArrayList<>();
        for(int i=0;i<nums.length-2;i++){
            if(i>0&&nums[i]==nums[i-1]){
                continue;
            }
            if(nums[i]>0) break;
            for(int j=i+1,k=nums.length-1;j<k;){
                int temp = nums[i]+nums[j]+nums[k];
                if(temp==0){
                    List<Integer> list = new ArrayList<>();
                    list.add(nums[i]);
                    list.add(nums[j]);
                    list.add(nums[k]);
                    result.add(list);
                    j++;
                    k--;
                    while(nums[j]==nums[j-1]&&j<k){
                        j++;
                    }
                    while(nums[k]==nums[k-1]&&j<k){
                        k--;
                    }
                }else if(temp<0){
                    j++;
                    while(nums[j]==nums[j-1]&&j<k){
                        j++;
                    }
                }else{
                    k--;
                    while(nums[k]==nums[k-1]&&j<k){
                        k--;
                    }
                }
            }
        }
        return result;
    }
}

File: src/main/java/五毒/第一周/第三遍/three_sum.go
package main

import (
	"sort"
)

func threeSum(nums []int) [][]int {
	sort.Ints(nums)
	result := [][]int{}
	for i := 0; i < len(nums)-2; i++ {
		if i > 0 && nums[i] == nums[i-1] {
			continue
		}
		l := i + 1
		r := len(nums) - 1
		for l < r {
			if nums[i]+nums[l]+nums[r] == 0 {
				item := []int{nums[i], nums[l], nums[r]}
				result = append(result, item)
				l++
				r--
				for l < r && nums[l] == nums[l-1] {
					l++
				}
				for l < r && nums[r] == nums[r+1] {
					r--
				}
			} else if nums[i]+nums[l]+nums[r] < 0 {
				l++
			} else {
				r--
			}
		}
	}
	return result
}

func main() {
	nums := []int{-1, 0, 1, 2, -1, -4}
	print(threeSum(nums))
}
